feat: Change ObjectStorePath API to be more explicit

carols10cents opened the pull request:
Now you have to designate whether you're adding a directory or a file
name, with some assumptions based on paths coming from a cloud object
storage or the file system.

A notable difference: checking to see if "apple/b" is a prefix of
"apple/bear/cow.json" will now say no; only whole directories are
matched.

pauldix commented:
@carols10cents what is the prefix matching for? I ask because we'll likely need to (at some point) go through the files in a dir but starting at some prefix.

carols10cents commented:
> @carols10cents what is the prefix matching for? I ask because we'll likely need to (at some point) go through the files in a dir but starting at some prefix.

It's currently being used to [implement the in memory `list` and `list_with_delimiter`](https://github.com/influxdata/influxdb_iox/pull/629/files#diff-db1903cfafe0a405a23d70e0a4139187a51c1db17e21258e74c8a09895c7c9ecL86-R142), which... errr... I thought was passing tests but we're still ignoring it 🤦‍♀️ ummm changing this to draft....

If it's files in a dir (so the prefix is the full directory name), that should work. What won't work right now is if you specify a partial directory name-- is that a problem? It shouldn't be bad to fix, it just didn't seem needed with what the code is doing right now.

pauldix commented:
@carols10cents the case we need to handle is if we have files like this `1/mydb/partitions/2021-01-09/00001/foo.parquet`. And say we have thousands of those parquet files in that dir. We need to be able to list from `1/mydb/partitions/2021-01-09/00001/m` onward. I guess this doesn't matter for the in-memory since it won't limit its results to 1,000 entries. Maybe it also doesn't matter for files? But I'm pretty sure we'll need that for the object stores.

carols10cents commented:
Ok, this is now ready for review!

The tests for the in-memory object store are now all unignored and passing. @pauldix please especially check the additional scenario I added in the `list_with_delimiter` test function to make sure I captured what we were discussing above accurately!

I had hoped to get the filesystem object store working today too, but that's proving... fun... so I'm going to do that in a follow-on PR instead.

alamb commented on object_store/src/path.rs:
```diff
@@ -13,86 +13,221 @@ use std::path::PathBuf;
+                self_parts == *other_parts
+            }
+            _ => {
+                let self_parts: DirsAndFileName = self.to_owned().into();
```
Am I reading this right that if you have a `Path::RawBuf` or a `Path::RawCloud` and you compare them with any other path, they will be converted to `DirsAndFileName` (which presumably is expensive otherwise we would do it always) but then the converted version is discarded.

So performance wise, if a Path is *ever* compared more than once it would have been better to convert to `DirsAndFileName` as part of creation?

carols10cents replied:
Yes, you're reading this right. The tricky part is that it would be strange if comparing paths mutated them.

I don't think comparing cloud/file paths to `DirsAndFileName` paths is likely to happen in production, only in tests: when we have cloud/file paths, we're going to be passing them to the cloud service for it to perform operations on, or to the file implementation to be used as `PathBuf`s.

alamb commented on object_store/src/path.rs:
```diff
@@ -124,7 +360,27 @@ impl CloudConverter {
+        match &object_store_path.inner {
+            PathRepresentation::RawCloud(path) => path.to_owned(),
+            PathRepresentation::RawPathBuf(_path) => {
+                todo!("convert");
```
I wonder if this should return an error or something other than panic'ing?

alamb commented on object_store/src/path.rs:
```diff
@@ -138,7 +394,23 @@ impl FileConverter {
-        object_store_path.parts.iter().map(|p| &p.0).collect()
+        match &object_store_path.inner {
+            PathRepresentation::RawCloud(_path) => {
+                todo!("convert");
```
likewise here

alamb reviewed (commented):
I have some stylistic comments about the value of deferring `DirsAndFileName` conversion, but I don't feel strongly about it.

I think the following two items would make this PR much stronger:

# `todo!()`
I think the `todo!()`s in the code should should be removed (or commented explaining why they are there)

# scenario test listing files
I would also suggest a test demonstrating how to list files in the usecase that @pauldix  mentioned, which I think is as follows.  I may have missed this test or perhaps it is coming in a future PR, but I was imagining something a scenario the following:

Setup: Paths in the object store (note they are sorted lexicographically):
```
1/mydb/partitions/2021-01-09/00001/foo.parquet
1/mydb/partitions/2021-01-09/00002/foo.parquet
1/mydb/partitions/2021-01-09/00003/foo.parquet
1/mydb/partitions/2021-01-09/00004/foo.parquet
1/mydb/partitions/2021-01-09/00005/foo.parquet
1/mydb/partitions/2021-01-09/00010/foo.parquet
1/mydb/partitions/2021-01-09/00010/bar.parquet # note different filename
1/mydb/partitions/2021-01-09/00011/foo.parquet
```

There is one operation to list "prefixes" starting at some prefix:

`list_prefixes_starting_at('1/mydb/partitions/2021-01-09/00005')` should return only the prefixes

```
1/mydb/partitions/2021-01-09/00005
1/mydb/partitions/2021-01-09/00010
1/mydb/partitions/2021-01-09/00011
```

And an equivalent that lists all files

`list_starting_at('1/mydb/partitions/2021-01-09/00005')` should return

```
1/mydb/partitions/2021-01-09/00005/foo.parquet
1/mydb/partitions/2021-01-09/00010/foo.parquet
1/mydb/partitions/2021-01-09/00010/bar.parquet # note different file naem
1/mydb/partitions/2021-01-09/00011/foo.parquet
```

carols10cents commented:
Ok, I think I see where I was getting confused... there are 2 ways to think about "starting with" a prefix :) 

One is that a path contains a prefix, so `1/mydb/partitions/2021-01-09/00010/foo.parquet` starts with the prefix `1/mydb/partitions/2021-01-09/00010/f`.

One is that an alphabetical list of paths can start *from* a point in the middle of the list, which is your example @alamb, a list of the objects in `1/mydb/partitions/2021-01-09/` starting at `1/mydb/partitions/2021-01-09/00005` will contain `1/mydb/partitions/2021-01-09/00010`, and `1/mydb/partitions/2021-01-09/00005` IS NOT a prefix of `1/mydb/partitions/2021-01-09/00010`.

This seems like 2 separate operations, but we only have `list` and `list_with_delimiter` right now, that take a `prefix` but not a "list starter".

I don't think the cloud services support this idea of a "list starter"; you can specify a prefix and a delimiter and get the common prefixes and the objects and paginate through the objects using the continuation token, not like "start from `m`".

Writing this up made me realize the `prefix_matches` method and thus the in memory object store was inconsistent with the cloud object stores; I've added more test cases and fixed that problem.

I think if I'm still misunderstanding the functionality you two want, @alamb and @pauldix, it would be super helpful if you could write up some failing test cases.

I do agree with your comments about the `todo`s; will take care of those now!

alamb reviewed (approved):
I personally suggest we merge this PR in (as it is an improvement over what we have) and then as @pauldix  works out his exact needs in the recovery / catalog metadata design we can add more features as necessary

Thanks for all the work @carols10cents !

carols10cents commented:
Ok, merging, especially because I think I've now convinced myself that you should only be able to convert cloud paths to DirAndFile and PathBuf to DirAndFile but not cloud -> PathBuf or vice versa, which means refactoring the internals again lol.
